scratchwork/fsdp_min/fsdp.py

```diff
@@ -21,51 +22,72 @@
 
 
 def get_fake_data_iter(
-    batch_size: int, vocab_size: int, max_seq_len: int, rank: int, device: torch.device
+    batch_size: int,
+    vocab_size: int,
+    max_seq_len: int,
+    rank: int,
+    device: torch.device,
+    is_validation: bool,
 ) -> Generator[tuple[torch.Tensor, torch.Tensor], None, None]:
     """
-    Fake dataloader. Repeatedly yields the same (inputs, targets) tuple of tensors, with different
-    tensors on different ranks.
+    Fake dataloader. Yields a different set of data for each rank, and for train vs validation.
+    This data would usually come from a tokenized dataset.
     """
-    torch.manual_seed(42 + rank)
-    fake_sequence = torch.randint(vocab_size, (batch_size, max_seq_len), device=device)
-    inputs, targets = fake_sequence[..., :-1], fake_sequence[..., 1:]
+    generator = torch.Generator(device=device)
+    generator.manual_seed(42 + rank + 100000 * is_validation)
     while True:
+        fake_sequence = torch.randint(
+            vocab_size, (batch_size, max_seq_len), device=device, generator=generator
```

Comment:
I think I had an error here: shape should be `(batch_size, max_seq_len + 1)` so that `inputs` and `targets` are shape `(batch_size, max_seq_len)` after slicing.

Reply:
Good catch; fixing.